tests/tracking/test_fluent.py

```diff
@@ -69,3 +76,266 @@ def test_get_experiment_id():
         assert exp_id is not None
         mlflow.set_experiment(name)
         assert _get_experiment_id() == exp_id
+
+
+@pytest.fixture
+def empty_active_run_stack():
+    with mock.patch("mlflow.tracking.fluent._active_run_stack", []):
+        yield
+
+
+def is_from_run(active_run, run):
+    return active_run.info == run.info and active_run.data == run.data
+
+
+def test_start_run_defaults(empty_active_run_stack):
+
+    mock_experiment_id = mock.Mock()
```

Comment:
These tests do confirm that we call the appropriate functions and add appropriate labels/tags, but won't they all break after you refactor the underling implementation such that, for example, `mlflow.tracking.fluent._get_source_name`, but may rather be embedded in some sort of DefaultRunContextProvider?

If we tested the apparent outcome (e.g., that source_name should be 'test_fluent'), then that would be robust to internal refactorings.

Reply:
Here's a sketch of the databricks test updated with this approach:
https://github.com/facultyai/mlflow/compare/fluent-start-run-tests...aarondav:fluent-start-run-tests

Upside: this test would demonstrate that your proposed refactoring in fact continues to work. Otherwise, it's possible that for example you copy `mlflow.tracking.fluent.get_notebook_path`, forget to delete the old code path, and still hae the test pass.

Downside: this becomes less of a unit test and more of a partial integration test, as it has to reach into databricks_utils.py's internal structure. I think this is somewhat reasonable since we're trying to avoid breaking the existing behavior, but ideally it would've been factored out into a ContextProvider interface in the first place so we don't have to special-case the test (but that's what we're moving towards anyway).

Reply:
Thanks for the comments - what you say about being robust to change is definitely relevant, though I have in mind to update these tests alongside subsequent refactoring in a way that provided confidence that the existing logic is retained.

What I have in mind is:

1. Move the helpers mocked in these tests (e.g. `_get_source_name`) to `mlflow.tracking.context` alongside the new `ContextProvider` logic and update the tests added in this PR to mock the new locations.
2. Verify that the tests added in this PR still pass.
3. Implement the context providers and use them in `mlflow.tracking.fluent.start_run`.
4. Verify that the tests added in the PR still pass - this will then check that the integration of `start_run` with the new context provider logic results in the same mapping of `start_run` inputs to `MlflowClient.create_run` inputs as currently.
5. With the refactoring complete, these tests can be revised to match the new code structure better. For example, some of the helpers might get merged into methods of context providers and we might want to test them directly and use mock context providers to test the new logic in `start_run`.

In general, I view the tests in this PR in part as a temporary scaffold so I don't bring the house down while refactoring.

What do you think?